File: kiit-codes/src/commonTest/kotlin/kiit/codes/StatusTest.kt
package kiit.codes

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertNotSame
import kotlin.test.assertSame
import kotlin.test.assertTrue

// =================================================================================================
// StatusTest: Passed/Failed subtypes, ofStatus companion function
// =================================================================================================

class StatusTest {
    // -------------------------------------------------------------------------
    // success flag: Passed subtypes (hoisted onto Passed itself, see Passed.success)
    // -------------------------------------------------------------------------

    @Test fun succeededHasSuccessTrue() {
        assertTrue(Passed.Succeeded("S", "S").success)
    }

    @Test fun pendingHasSuccessTrue() {
        assertTrue(Passed.Pending("P", "P").success)
    }

    @Test fun excludedHasSuccessTrue() {
        assertTrue(Passed.Excluded("F", "F").success)
    }

    @Test fun informationHasSuccessTrue() {
        assertTrue(Passed.Information("I", "I").success)
    }

    // -------------------------------------------------------------------------
    // success flag: Failed subtypes (hoisted onto Failed itself, see Failed.success)
    // -------------------------------------------------------------------------

    @Test fun restrictedHasSuccessFalse() {
        assertFalse(Failed.Restricted("R", "R").success)
    }

    @Test fun invalidHasSuccessFalse() {
        assertFalse(Failed.Invalid("I", "I").success)
    }

    @Test fun rejectedHasSuccessFalse() {
        assertFalse(Failed.Rejected("E", "E").success)
    }

    @Test fun unservedHasSuccessFalse() {
        assertFalse(Failed.Unserved("U", "U").success)
    }

    // -------------------------------------------------------------------------
    // origin: defaults to "custom" for direct construction, overridable at the call site
    // -------------------------------------------------------------------------

    @Test fun originDefaultsToCustom() {
        assertEquals(StatusConstants.CUSTOM, Passed.Succeeded("S", "S").origin)
        assertEquals(StatusConstants.CUSTOM, Failed.Restricted("R", "R").origin)
    }

    @Test fun originIsOverridable() {
        assertEquals(StatusConstants.KIIT, Passed.Succeeded("S", "S", origin = StatusConstants.KIIT).origin)
    }

    // -------------------------------------------------------------------------
    // group: the discriminant field, exhaustive over all 8 subtypes
    // -------------------------------------------------------------------------

    @Test
    fun groupReturnsCorrectStringForAllSubtypes() {
        assertEquals("Succeeded", Passed.Succeeded("S", "S").group)
        assertEquals("Pending", Passed.Pending("P", "P").group)
        assertEquals("Excluded", Passed.Excluded("F", "F").group)
        assertEquals("Information", Passed.Information("N", "N").group)
        assertEquals("Restricted", Failed.Restricted("R", "R").group)
        assertEquals("Invalid", Failed.Invalid("I", "I").group)
        assertEquals("Rejected", Failed.Rejected("E", "E").group)
        assertEquals("Unserved", Failed.Unserved("U", "U").group)
    }

    // -------------------------------------------------------------------------
    // groupDescription: runtime-accessible version of each group's meaning
    // -------------------------------------------------------------------------

    @Test
    fun groupDescriptionIsNonBlankAndDistinctForAllSubtypes() {
        val descriptions =
            listOf(
                Passed.Succeeded("S", "S").groupDescription,
                Passed.Pending("P", "P").groupDescription,
                Passed.Excluded("F", "F").groupDescription,
                Passed.Information("N", "N").groupDescription,
                Failed.Restricted("R", "R").groupDescription,
                Failed.Invalid("I", "I").groupDescription,
                Failed.Rejected("E", "E").groupDescription,
                Failed.Unserved("U", "U").groupDescription,
            )
        assertTrue(descriptions.all { it.isNotBlank() })
        assertEquals(descriptions.size, descriptions.toSet().size)
    }

    @Test
    fun groupDescriptionIsConsistentAcrossInstancesOfTheSameSubtype() {
        assertEquals(Failed.Restricted("A", "A").groupDescription, Failed.Restricted("B", "B").groupDescription)
    }

    // -------------------------------------------------------------------------
    // scope: defaulted field on every concrete Passed/Failed subtype, "" means unset
    // -------------------------------------------------------------------------

    @Test
    fun scopeDefaultsToEmptyString() {
        assertEquals("", Failed.Restricted("RESTRICTED", "Restricted").scope)
        assertEquals("", Succeeded.SUCCESS.scope)
    }

    @Test
    fun scopeIsSettableViaConstructor() {
        val s = Failed.Restricted("RESTRICTED", "Restricted", origin = StatusConstants.KIIT, scope = "payments.cards")
        assertEquals("payments.cards", s.scope)
    }

    @Test
    fun scopeIsSettableViaCopyOnAnExistingInstanceIncludingBuiltIns() {
        val scoped = Succeeded.CREATED.copy(scope = "payments.cards")
        assertEquals("payments.cards", scoped.scope)
        assertEquals(Succeeded.CREATED.name, scoped.name)
        assertEquals(Succeeded.CREATED.origin, scoped.origin)
    }

    // -------------------------------------------------------------------------
    // statusKey: module-internal StatusKey(origin, scope, group, name), used to key
    // CodesToHttp/CodesToGrpc overrides.
    // -------------------------------------------------------------------------

    @Test
    fun statusKeyHasEmptyScopeByDefault() {
        val s = Failed.Restricted("RESTRICTED", "Restricted", origin = StatusConstants.KIIT)
        assertEquals(StatusKey(StatusConstants.KIIT, "", "Restricted", "RESTRICTED"), s.statusKey)
    }

    @Test
    fun statusKeyIncludesScopeWhenSet() {
        val s = Failed.Restricted("RESTRICTED", "Restricted", origin = StatusConstants.KIIT, scope = "payments.cards")
        assertEquals(StatusKey(StatusConstants.KIIT, "payments.cards", "Restricted", "RESTRICTED"), s.statusKey)
    }

    @Test
    fun statusKeyDiffersByScopeForOtherwiseIdenticalStatuses() {
        val base = Failed.Restricted("RESTRICTED", "Restricted", origin = StatusConstants.KIIT)
        val scopedA = base.copy(scope = "payments.cards")
        val scopedB = base.copy(scope = "payments.wallets")
        assertNotEquals(scopedA.statusKey, scopedB.statusKey)
        assertNotEquals(base.statusKey, scopedA.statusKey)
    }

    // -------------------------------------------------------------------------
    // path: "$origin:$scope" when scope is set, else just $origin
    // -------------------------------------------------------------------------

    @Test
    fun pathIsJustOriginWhenScopeIsUnset() {
        val s = Failed.Restricted("RESTRICTED", "Restricted", origin = StatusConstants.KIIT)
        assertEquals(StatusConstants.KIIT, s.path)
    }

    @Test
    fun pathIncludesScopeWhenSet() {
        val s = Failed.Restricted("RESTRICTED", "Restricted", origin = StatusConstants.KIIT, scope = "payments.cards")
        assertEquals("${StatusConstants.KIIT}:payments.cards", s.path)
    }

    // -------------------------------------------------------------------------
    // code: "${Passed|Failed}:$group:$name", stable across scope, not unique
    // -------------------------------------------------------------------------

    @Test
    fun codeForPassedStatus() {
        assertEquals("Passed:Succeeded:SUCCESS", Succeeded.SUCCESS.code)
    }

    @Test
    fun codeForFailedStatus() {
        val s = Failed.Restricted("RESTRICTED", "Restricted", origin = StatusConstants.KIIT)
        assertEquals("Failed:Restricted:RESTRICTED", s.code)
    }

    @Test
    fun codeIsUnaffectedByScope() {
        val s = Succeeded.SUCCESS.copy(scope = "payments.cards")
        assertEquals("Passed:Succeeded:SUCCESS", s.code)
    }

    @Test
    fun codeIsNotUniqueAcrossDifferentOriginsOrScopes() {
        val kiitDenied = Failed.Restricted("DENIED", "Denied", origin = StatusConstants.KIIT)
        val customDenied = Failed.Restricted("DENIED", "Custom denied", origin = "com.acme")
        assertEquals(kiitDenied.code, customDenied.code)
        assertNotEquals(kiitDenied.statusKey, customDenied.statusKey)
    }

    // -------------------------------------------------------------------------
    // ofStatus: selects correct instance based on message / rawStatus nullability
    // -------------------------------------------------------------------------

    @Test
    fun ofStatusReturnStatusWhenBothNull() {
        val status = Succeeded.SUCCESS
        assertSame(status, Status.ofStatus(null, null, status))
    }

    @Test
    fun ofStatusReturnsRawStatusWhenMessageIsNull() {
        val raw = Succeeded.CREATED
        val result = Status.ofStatus(null, raw, Succeeded.SUCCESS)
        assertSame(raw, result)
    }

    @Test
    fun ofStatusReturnsStatusWithUpdatedMessageWhenRawIsNull() {
        val result = Status.ofStatus("Custom", null, Succeeded.SUCCESS)
        assertEquals("Custom", result.message)
        assertEquals(Succeeded.SUCCESS.origin, result.origin)
    }

    @Test
    fun ofStatusReturnsRawWithUpdatedMessageWhenBothProvided() {
        val raw = Succeeded.CREATED
        val result = Status.ofStatus("Custom", raw, Succeeded.SUCCESS)
        assertEquals("Custom", result.message)
        assertEquals(raw.origin, result.origin)
        assertNotSame(raw, result)
    }

    // -------------------------------------------------------------------------
    // Typealiases (Restricted, Invalid, ...) are fully transparent: the same type as their
    // Failed.X/Passed.X target, not a copy. A `when` mixing aliased and fully-qualified branches
    // must still be exhaustive to the compiler, this is as much a compile-time check as a
    // runtime one, it wouldn't build if the aliases introduced a distinct type.
    // -------------------------------------------------------------------------

    @Test
    fun aliasedAndFullyQualifiedBranchesAreExhaustiveInTheSameWhen() {
        val status: Status = Restricted.DENIED
        val label =
            when (status) {
                is Restricted -> "restricted"
                is Failed -> "failed"
                is Passed -> "passed"
            }
        assertEquals("restricted", label)
    }

    @Test
    fun aliasedConstantIsSameInstanceAsFullyQualifiedConstant() {
        assertSame(Failed.Restricted.DENIED, Restricted.DENIED)
    }
}

File: ports/kiit-codes-ts/test/status.test.ts
import { describe, expect, it } from "vitest";
import {
  Groups,
  StatusConstants,
  Succeeded,
  Pending,
  Excluded,
  Information,
  Restricted,
  Invalid,
  Rejected,
  Unserved,
} from "../src/groups.js";
import { groupDescription, statusPath, statusCode } from "../src/status.js";

// Ported from StatusTest.kt. Two Kotlin test groups are intentionally not ported:
// - statusKey checks: StatusKey is `internal` in Kotlin (module-private, inaccessible to real
//   external consumers too) and has no exported TS equivalent; covered indirectly via codes.test.ts's
//   CodesToHttp scope/override behavior instead.
// - typealias transparency checks (Restricted === Failed.Restricted, same instance): TS has no
//   typealias-vs-fully-qualified duality, groups.ts's Restricted is the only representation.

describe("Status: success flag", () => {
  it("is true for every Passed group", () => {
    expect(Succeeded("S", "S").success).toBe(true);
    expect(Pending("P", "P").success).toBe(true);
    expect(Excluded("F", "F").success).toBe(true);
    expect(Information("I", "I").success).toBe(true);
  });

  it("is false for every Failed group", () => {
    expect(Restricted("R", "R").success).toBe(false);
    expect(Invalid("I", "I").success).toBe(false);
    expect(Rejected("E", "E").success).toBe(false);
    expect(Unserved("U", "U").success).toBe(false);
  });
});

describe("Status: origin", () => {
  it("defaults to custom for direct construction", () => {
    expect(Succeeded("S", "S").origin).toBe(StatusConstants.CUSTOM);
    expect(Restricted("R", "R").origin).toBe(StatusConstants.CUSTOM);
  });

  it("is overridable at the call site", () => {
    expect(Succeeded("S", "S", StatusConstants.KIIT).origin).toBe(StatusConstants.KIIT);
  });
});

describe("Status: group", () => {
  it("returns the correct string for all eight groups", () => {
    expect(Succeeded("S", "S").group).toBe(Groups.SUCCEEDED);
    expect(Pending("P", "P").group).toBe(Groups.PENDING);
    expect(Excluded("F", "F").group).toBe(Groups.EXCLUDED);
    expect(Information("N", "N").group).toBe(Groups.INFORMATION);
    expect(Restricted("R", "R").group).toBe(Groups.RESTRICTED);
    expect(Invalid("I", "I").group).toBe(Groups.INVALID);
    expect(Rejected("E", "E").group).toBe(Groups.REJECTED);
    expect(Unserved("U", "U").group).toBe(Groups.UNSERVED);
  });
});

describe("groupDescription", () => {
  it("is non-blank and distinct for all eight groups", () => {
    const descriptions = [
      groupDescription(Succeeded("S", "S")),
      groupDescription(Pending("P", "P")),
      groupDescription(Excluded("F", "F")),
      groupDescription(Information("N", "N")),
      groupDescription(Restricted("R", "R")),
      groupDescription(Invalid("I", "I")),
      groupDescription(Rejected("E", "E")),
      groupDescription(Unserved("U", "U")),
    ];
    expect(descriptions.every((d) => d.trim().length > 0)).toBe(true);
    expect(new Set(descriptions).size).toBe(descriptions.length);
  });

  it("is consistent across instances of the same group", () => {
    expect(groupDescription(Restricted("A", "A"))).toBe(groupDescription(Restricted("B", "B")));
  });
});

describe("Status: scope", () => {
  it("defaults to an empty string", () => {
    expect(Restricted("RESTRICTED", "Restricted").scope).toBe("");
    expect(Succeeded.SUCCESS.scope).toBe("");
  });

  it("is settable via the constructor", () => {
    const s = Restricted("RESTRICTED", "Restricted", StatusConstants.KIIT, "payments.cards");
    expect(s.scope).toBe("payments.cards");
  });

  it("is settable via spread on an existing instance, including built-ins", () => {
    const scoped = { ...Succeeded.CREATED, scope: "payments.cards" };
    expect(scoped.scope).toBe("payments.cards");
    expect(scoped.name).toBe(Succeeded.CREATED.name);
    expect(scoped.origin).toBe(Succeeded.CREATED.origin);
  });
});

describe("statusPath", () => {
  it("is just origin when scope is unset", () => {
    const s = Restricted("RESTRICTED", "Restricted", StatusConstants.KIIT);
    expect(statusPath(s)).toBe(StatusConstants.KIIT);
  });

  it("includes scope when set", () => {
    const s = Restricted("RESTRICTED", "Restricted", StatusConstants.KIIT, "payments.cards");
    expect(statusPath(s)).toBe(`${StatusConstants.KIIT}:payments.cards`);
  });
});

describe("statusCode", () => {
  it("is Passed:group:name for a passed status", () => {
    expect(statusCode(Succeeded.SUCCESS)).toBe("Passed:Succeeded:SUCCESS");
  });

  it("is Failed:group:name for a failed status", () => {
    const s = Restricted("RESTRICTED", "Restricted", StatusConstants.KIIT);
    expect(statusCode(s)).toBe("Failed:Restricted:RESTRICTED");
  });

  it("is unaffected by scope", () => {
    const s = { ...Succeeded.SUCCESS, scope: "payments.cards" };
    expect(statusCode(s)).toBe("Passed:Succeeded:SUCCESS");
  });

  it("is not unique across different origins", () => {
    const kiitDenied = Restricted("DENIED", "Denied", StatusConstants.KIIT);
    const customDenied = Restricted("DENIED", "Custom denied", "com.acme");
    expect(statusCode(kiitDenied)).toBe(statusCode(customDenied));
  });
});
